spec/index.md

```diff
@@ -2114,6 +2114,16 @@ This function returns fee percentiles, measured in millisatoshi/vbyte (1000 mill
 
 The [standard nearest-rank estimation method](https://en.wikipedia.org/wiki/Percentile#The_nearest-rank_method), inclusive, with the addition of a 0th percentile is used. Concretely, for any i from 1 to 100, the ith percentile is the fee with rank `⌈i * 100⌉`. The 0th percentile is defined as the smallest fee (excluding coinbase transactions).
 
+### IC method `metrics` {#ic-metrics}
+
+Given a subnet ID as input, this method returns a collection of metrics for the given subnet.
+
+A single metric entry is of an enumeration type with a single variant (in the future, more variants might be added):
+
+- `node_metrics`: provides metrics for a node characterized by its principal (field `node_id`), for a period of time characterized by start and end timestamps in nanoseconds since 1970-01-01 (fields `start_timestamp_nanos` and `end_timestamp_nanos`), and the actual metrics values:
```

Comment:
Having a `start_timestamp` and `end_timestamp` attached to a bunch of gauges is not ideal.

On the one hand, it limits the resolution of the metrics. Although in this case that limitation comes mostly from the checkpointing interval or query epoch. But if we were to change the implementation of either of those in the future and be able to update said metric more often (e.g. every round), we'd either have to retain the artificially reduced resolution or have `start_timestamp` and `end_timestamp` one second apart.

Which leads to the second issue: if anyone only cares about the cumulative amount over a longer period (e.g. how many queries were executed by the subnet over the past day) they will be forced to query the metrics at least every checkpoint interval (or possibly more often) and deduplicate and maybe even interpolate to get a reasonable value.

A third issue is that as you add more and more metrics (and they are collected with different resolutions: e.g. number of replicated calls is updated every round; node metrics every checkpoint interval; query metrics every "query epoch") you'll end up with different metrics covering different and even non-overlapping time ranges (e.g. node metrics would be from 10 minutes ago; query metrics from over half an hour ago).

The way Prometheus works around this (and we also have such subnet metrics ourselves) is to use a monotonically increasing accumulator (i.e. a [counter](https://prometheus.io/docs/tutorials/understanding_metric_types/)). We could label it as e.g. "total queries executed by subnet", but the absolute value is usually not that relevant (because it depends on when the subnet was created or when the metric started being recorded). OTOH, if someone collects at least a couple of samples of this metric at different times, they can trivially compute the difference in observations between the first and the second "scrape". One can scrape the metrics endpoint as often or as rarely as they want and they will still get meaningful data out of it.

I.e. I would strongly suggest using a counter instead of gauges over arbitrary and possibly disjoint time ranges.

Reply:
I think that a simple counter would not be enough based on how I understand the requirements. The goal of this feature is to be able to have a way to query the entire metrics for (at least) the previous month with a resolution of one day. To achieve this with a single counter would not work.

Maybe the requirements can somehow be adapted to make a single counter sufficient? A single counter would arguably be much simpler in many respects.

Reply:
Thanks for your comments!
I agree that a counter with a timestamp of its last change would be much simpler.
 In the future (after IC-1591), I'd envision automated NP remuneration based on such a timestamped counter, executed by a NNS-controlled canister. In the short term however, (i.e., the [feature IC-1591 we're working on right now](https://docs.google.com/presentation/d/1KXIZtYtn658L2mDKrgmL502jeBo1QbHoJRYYAg7Pu68/edit#slide=id.p)), we want to allow users to have access to historical data on node performance, ideally without building additional canister logic.
 
In a discussion last week (with Sasa, Martin, David and Dimitris), we discussed different options (listed in the slide set linked above), with and without storing historical data and making it available via this interface. 
 
 
Option 1: 

* store num_blocks and num_block_failure counters per node in subnet, 
* management canister metrics method returns list of all nodes’ current counter values together with timestamp
* build a canister storing in historical values by making  periodic calls to metrics method, stores them and makes them available via canisters methods 

+ Very simple implementation on the replica side
- Development and maintenance of new canister logic: currently there is no team responsible for such a canister, knowhow and ownership would need to be built up
- Auditability requires analyzing the canister that stores the data and its changes


Option 2:
* store num_blocks and num_block_failure counters per day (using data from first block of new calendar day) per node in subnet for the last 60 days 
* management canister metrics method returns list of all nodes’ counter values for a given day

+ Very simple auditability for NPs and IC community
+ No new canisters to be developed and maintained
- Slightly more complexity to prepare stored data and return the metrics for the desired timestamp
- More data needs to be stored (could be reduced to keeping 2 old counter values + timestamps, if 60 is too much)

In order to allow NPs and the IC community to see how the nodes perform soon and more easily, I currently favour Option 2 (less DFINITY effort, easier for users). Note that in any case this API change will be marked experimental and is expected to change.

Reply:
Agreed. The start timestamp plus end timestamp made it look like these were gauges ("this many blocks between start timestamp and end timestamp"), which may or may not have been the intent.

Retaining the last N values of a counter at specific timestamps (at intervals of 1 day or not), while not ideal, is certainly a lot cleaner and opens he possibility of later going for a single sample at the "current timestamp" without even changing the API.

Reply:
We ended up with a compromise between Option 1 and Option 2, so I'm closing this thread.